### DISCUSSION TITLE
Format 2 different variables with ChooseFormatter

### DISCUSSION THREAD

**Discussion by mhspelt** [axuno/SmartFormat#198] [Q&A] [answered]
Hi,
I have an object with a nullable and non-nullable field of the same type. I'd like to format the result of the null-coalesced value but I cannot figure out how to do it. 

**{data.NullableNumber:N0}** works if NullableNumber is set.
**{data.NullableNumber:choose(null): data.NonNullableNumber | {}}** works but shows the number without formatting

I want to combine these options and tried something like

{data.NullableNumber:N0:choose(null): {data.NonNullableNumber:N0} | {}}

or 

{{data.NullableNumber:choose(null): data.NonNullableNumber | {}}:N0}

but parsing fails for all of them. I guess what I want is simply

{data.NullableNumber ?? data.NonNullableNumber:N0}

**Comment by axunonb** [comment, marked as answer]
Hi @mhspelt,
Assuming you're on v2.7.0, would you agree that this works as expected?
```CSharp
var data = new { NullableNumber = (int?) 9876, NonNullable = (int) 1234};
var result = Smart.Format(CultureInfo.InvariantCulture, "{NullableNumber:choose(null):{NonNullable:N2}|{:N2}}", data);
// result: 9,876.00
data = new { NullableNumber = (int?) null, NonNullable = (int) 1234};
result = Smart.Format(CultureInfo.InvariantCulture, "{NullableNumber:choose(null):{NonNullable:N2}|{:N2}}", data);
// result: 1,234.00
```
Number format may differ depending on culture setting.

**Reply by mhspelt** [reply]
Sorry to reopen, but I tried the following template with a nullable date field and it did not work:

`{nullableDateField:choose(null): null|{Date:dd-MM}}`

I'm just getting the input back. If I try only

`{nullableDateField:choose(null): null|{Date}}`

without the formatter, I get "null" or the unformatted date ("12/10/2021 00:00:00").

**Reply by axunonb** [reply]
It will work using the syntax from the [last answer](https://github.com/axuno/SmartFormat/discussions/198#discussioncomment-1291757)
```C#
var data = new { NullableDate = (DateTime?) new DateTime(2021,10,12), NonNullable = new DateTime(2021,1,1)};
var result = Smart.Format(CultureInfo.InvariantCulture, "{NullableDate:choose(null):{NonNullable:dd-MM}|{:dd-MM}}", data);
// result: 12-10
data = new { NullableDate = (DateTime?) null, NonNullable = new DateTime(2021,1,1)};
result = Smart.Format(CultureInfo.InvariantCulture, "{NullableDate:choose(null):{NonNullable:dd-MM}|{:dd-MM}}", data);
// result: 01-01
```

**Reply by mhspelt** [reply]
Ah, I found why it is not working: the double closing brace is escaped:
```
The format string has 1 issue:
Format string is missing a closing brace
In: "{object.DateField:choose(null): null|{:dd-MM}}"
At:  --------------------------------------------^ 
```

If I include an extra space 
```
"{NullableDate:choose(null):{NonNullable:dd-MM}|{:dd-MM} }", data);
--------------------------------------------------------^ here
```

it works.

**Reply by axunonb** [reply]
I'm so much working on v3 that I forgot this issue in v2, you're right.

**Reply by mhspelt** [reply]
FYI, I found `Smart.Default.Parser.UseAlternativeEscapeChar()` as a workaround.
